src/test/java/com/xwiki/authentication/saml/XWikiGroupManagerTest.java

```diff
@@ -0,0 +1,164 @@
+import org.xwiki.model.reference.DocumentReferenceResolver;
+
+import javax.servlet.http.HttpSession;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class XWikiGroupManagerTest {
+    @Test
+    public void addUserToGroup_ShouldAddUserToGivenGroup() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+        .whenAddedToGroup()
+        .then()
+            .user("ArthurDent")
+            .isInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void removeUserFromGroup_ShouldRemoveUserFromGivenGroup() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+            .userIsAlreadyInGroup("ArthurDent","StarshipTroopers")
+        .whenRemoveFromGroup()
+        .then()
+            .user("ArthurDent")
+            .isntInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void userAlreadyMemberOfGroup_ShouldIgnoreTheNewMembership() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+            .userIsAlreadyInGroup("ArthurDent","StarshipTroopers")
+        .whenAddedToGroup()
+        .then()
+            .user("ArthurDent")
+            .isInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void removeUserFromGroupWhereHeIsnt_ShouldNotChangeGroupMembership() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+        .whenRemoveFromGroup()
+        .then()
+            .user("ArthurDent")
+            .isntInGroup("StarshipTroopers");
+    }
+
+    private DSL given(){
+        return new DSL();
+    }
+
+    private static class DSL {
+        private final XWikiContext context = new XWikiContext();
+        private XWikiGroupManager groupManager;
+
+        private final DocumentReferenceResolver<String> currentMixedDocumentReferenceResolver =
+                (docRef, parameters) -> new DocumentReference("XWiki", "XWikiGroups", docRef);
+
+        private String userName;
+        private String groupName;
+
+        final XWikiMock xwiki;
+        Logger root = (Logger) LoggerFactory.getLogger(org.slf4j.Logger.ROOT_LOGGER_NAME);
+
+        public DSL() {
+            root.setLevel(Level.OFF);
+            context.setWikiId("XWIKI");
+            context.setMainXWiki("MAIN-XWIKI");
+            context.setAction("");
+            context.setResponse(mock(XWikiResponse.class));
+            XWikiRequest request = mock(XWikiRequest.class);
+            HttpSession httpSession = mock(HttpSession.class);
+            when(request.getSession(true)).thenReturn(httpSession);
+            when(request.getSession()).thenReturn(httpSession);
+
+            context.setRequest(request);
+            xwiki = new XWikiMock(context);
+            context.setWiki(xwiki);
+        }
```

Comment:
If you rewrite this code like this:

```java
        public DSL() {
            root.setLevel(Level.OFF);
            
            // Mocking request
            XWikiRequest request = mock(XWikiRequest.class);
            HttpSession httpSession = mock(HttpSession.class);
            when(request.getSession(true)).thenReturn(httpSession);
            when(request.getSession()).thenReturn(httpSession);
            
            // Mocking XWiki
            xwiki = new XWikiMock(context);

            // Setting up context
            context.setWikiId("XWIKI");
            context.setMainXWiki("MAIN-XWIKI");
            context.setAction("");
            context.setResponse(mock(XWikiResponse.class));
            context.setRequest(request);
            context.setWiki(xwiki);
        }
```

There is a more clear separation of what you are doing, may even be able to extract some methods...